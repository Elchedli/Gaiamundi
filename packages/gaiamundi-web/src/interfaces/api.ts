export interface ApiError {
  status?: number;
  name?: string;
  message?: string;
  description?: string;
  details?: {
    errors?: Array<{
      message: string;
      name: string;
      path: string[];
    }>;
  };
}

export interface ApiErrorResponse {
  data?: any;
  error?: ApiError;
}

export type ApiResponse<T> = T;

export type ApiPaginationMeta = {
  //
  start?: number;
  limit?: number;
  // OR
  page?: number;
  pageSize?: number;
  pageCount?: number;
  // Always
  total: number;
};

export type ApiData<T> = T & {
  id: number;
  created_at: string;
  updated_at: string;
};

export interface ApiDocument<T> {
  data: ApiData<T>;
  meta?: {
    availableLocales?: string[];
  };
}

export interface ApiCollection<T> {
  data: Array<ApiData<T>>;
  meta: {
    pagination: ApiPaginationMeta;
  };
}
