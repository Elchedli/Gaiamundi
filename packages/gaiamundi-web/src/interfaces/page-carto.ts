import { ApiData, ApiDocument } from './api';
import { GeoMapAttributes } from './geo-map';
import { User } from './user';

export interface Tags {
  id: number;
  name: string;
  type: string[];
}

export interface PageCartoAttributes {
  id?: string;
  name: string;
  owner?: ApiDocument<User>;
  map: GeoMapAttributes;
  html: HTMLElement;
  tags: Tags;
  cover?: MediaImage;
}

export type PageCarto = ApiData<PageCartoAttributes>;
