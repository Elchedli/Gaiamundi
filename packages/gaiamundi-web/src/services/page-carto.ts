import { PageCarto, PageCartoStub } from 'interfaces/page-carto';
import { ContentType, strapi } from './strapi';

export const createPageCarto = async (data: PageCartoStub) => {
  return await strapi.create<PageCartoStub>(ContentType.PAGE_CARTOS, data);
};

export const getLatestPageCartos = async (page: number, pageSize: number) => {
  return await strapi.get<PageCarto>(ContentType.PAGE_CARTOS, {
    populate: '*',
    sort: 'createdAt:desc',
    pagination: {
      page,
      pageSize,
    },
  });
};

export const getPageCartoById = async (id: number) => {
  return await strapi.getById<PageCarto>(ContentType.PAGE_CARTOS, id, {
    populate: {
      map: {
        populate: {
          geoJSON: true,
        },
      },
      owner: true,
      columns: {
        populate: {
          dataset: true,
        },
      },
    },
  });
};

export const uploadCsv = async (file: File) => {
  return await strapi.uploadFile(file, 'api::page-carto.page-carto');
};
