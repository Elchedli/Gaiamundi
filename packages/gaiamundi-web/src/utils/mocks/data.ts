import { ApiCollection, ApiData } from 'interfaces/api';
import { DataFragment } from 'interfaces/data-fragment';
import { Dataset } from 'interfaces/dataset';
import { UploadedFile } from 'interfaces/file';
import { GeoMap } from 'interfaces/geo-map';
import { Indicator } from 'interfaces/indicator';
import { PageCarto } from 'interfaces/page-carto';
import { Tag } from 'interfaces/tag';
import { User } from 'interfaces/user';
import mockMapJson from './map.json';

export const mockMapPath = mockMapJson;

export const mockFileUpload = {
  id: 184,
  name: 'exampleParsing2.csv',
  alternativeText: null,
  caption: null,
  width: null,
  height: null,
  formats: null,
  hash: 'example_Parsing2_ce8756a044',
  ext: '.csv',
  mime: 'text/csv',
  size: 0.05,
  url: '/uploads/example_Parsing2_ce8756a044.csv',
  previewUrl: null,
  provider: 'local',
  provider_metadata: null,
  createdAt: '2023-03-16T18:20:09.476Z',
  updatedAt: '2023-03-16T18:20:09.476Z',
};

export const mockUser: User = {
  id: 1,
  username: 'testuser',
  password: 'test',
  email: 'test@test.com',
  provider: 'local',
  confirmed: true,
  blocked: false,
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoJson: ApiData<UploadedFile> = {
  id: 1,
  name: 'test.geojson',
  hash: 'abc123',
  provider: 'provider123',
  ext: '.geojson',
  mime: 'application/json',
  size: 1234,
  url: '/api/v1/geo-maps/id/thumbnail',
  formats: {
    thumbnail: {
      extension: 'ext',
      url: 'happyUrl',
      size: 360,
      hash: 'hashbar',
      mime: 'mimeExample',
    },
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockCover: ApiData<UploadedFile> = {
  id: 3,
  name: 'image.png',
  hash: 'hashImage',
  provider: 'provider123',
  ext: '.png',
  mime: 'application/image',
  size: 1234,
  url: '/api/v1/geo-maps/id/thumbnailImage',
  formats: {
    small: {
      extension: 'ext',
      url: '/api/v1/geo-maps/id/thumbnailImage?720',
      size: 360,
      hash: 'hash720',
      mime: 'mimeExample',
    },
    thumbnail: {
      extension: 'ext',
      url: '/api/v1/geo-maps/id/thumbnailImage?720',
      size: 360,
      hash: 'hash720',
      mime: 'mimeExample',
    },
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoMap: ApiData<GeoMap> = {
  id: 1,
  name: 'Test Map',
  yearValidity: 2022,
  mesh: 'Test Maille',
  source: 'Test Source',
  license: 'Test License',
  properties: [
    {
      name: 'code',
      sample: '39',
      isGeoCode: true,
    },
    {
      name: 'nom',
      sample: 'Jura',
      isGeoCode: false,
    },
  ],
  owner: mockUser,
  geoJSON: mockGeoJson,
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockTags: ApiData<Tag>[] = [
  {
    id: 1,
    name: 'Tag A',
    count: 5,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  {
    id: 2,
    name: 'Tag B',
    count: 3,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  {
    id: 3,
    name: 'Tag C',
    count: 2,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
];

export const mockDataset: ApiData<Dataset> = {
  id: 122,
  name: 'firstDataset',
  origin: 3,
  isPublic: true,
  owner: mockUser,
  csv: {
    id: 5,
    name: 'file.csv',
    hash: 'hashCSV',
    provider: 'provider123',
    ext: '.csv',
    mime: 'application/csv',
    size: 1234,
    url: '/api/v1/Uploads/id/csvFile.csv',
    formats: {
      thumbnail: {
        extension: 'ext',
        url: '/api/v1/Uploads/id/csvFile.csv?720',
        size: 360,
        hash: 'hash720',
        mime: 'mimeExample',
      },
    },
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockDataFragments: ApiData<DataFragment>[] = [
  {
    id: 1,
    name: 'FirstFragment',
    columns: [
      {
        name: 'firstColumn',
        sample: 5,
        source: 'source1',
        validity: '2021',
        isGeoCode: true,
      },
      {
        name: 'secondColumn',
        sample: 10.3,
        source: 'source2',
        validity: '2022',
        isGeoCode: false,
      },
      {
        name: 'thirdColumn',
        sample: 0.25,
        source: 'source3',
        validity: '2023',
        isGeoCode: false,
      },
      {
        name: 'fourthColumn',
        sample: 1000,
        source: 'source4',
        validity: '2024',
        isGeoCode: false,
      },
    ],
    dataset: mockDataset,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
];

export const indicatorsMockData: ApiData<Indicator>[] = [
  {
    id: 4,
    name: 'mockIndicator1',
    description: 'This is a mock indicator',
    source: 'mock',
    validity: 2,
    created_at: '2023-03-14T10:06:41.979Z',
    updated_at: '2023-03-14T10:06:41.979Z',
    equation: 'A+2',
    variables: [],
  },
  {
    id: 5,
    name: 'mockIndicator2',
    description: 'This is a mock indicator',
    source: 'mock',
    validity: 4,
    variables: [],
    created_at: '2023-03-14T10:12:37.285Z',
    updated_at: '2023-03-14T10:12:37.285Z',
    equation: 'A*2',
  },
];

export const mockPageCartoData: ApiData<PageCarto> = {
  id: 1,
  name: 'Page Carto 1',
  owner: mockUser,
  tags: mockTags,
  map: mockGeoMap,
  cover: mockCover,
  indicators: indicatorsMockData,
  charts: [],
  html: '<p>Test HTML</p>',
  data_fragments: mockDataFragments,
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoMapData: ApiData<GeoMap> = {
  id: 1,
  name: 'Test Map',
  yearValidity: 2022,
  mesh: 'Test Maille',
  source: 'Test Source',
  license: 'Test License',
  properties: [
    {
      name: 'nom',
      sample: 'Taponas',
      isGeoCode: true,
    },
  ],
  owner: mockUser,
  geoJSON: {
    id: 1,
    name: 'test.geojson',
    hash: 'abc123',
    provider: 'provider123',
    ext: '.geojson',
    mime: 'application/json',
    size: 1234,
    url: 'https://example.com/api/v1/geo-maps/id/thumbnail',
    formats: {
      '720': {
        extension: 'ext',
        url: 'happyUrl',
        size: 360,
        hash: 'hashbar',
        mime: 'mimeExample',
      },
    },
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoMapApiCollection: ApiCollection<GeoMap> = {
  data: Array.from({ length: 3 }, (_, index) => ({
    ...mockGeoMapData,
    id: index + 1,
    name: `Test Map ${index + 1}`,
  })),
  meta: {
    pagination: {
      pageCount: 1,
      total: 3,
    },
  },
};
