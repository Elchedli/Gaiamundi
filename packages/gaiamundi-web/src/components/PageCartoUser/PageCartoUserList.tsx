import { useQuery } from 'react-query';
import { useEffect, useState } from 'react';
import { ApiErrorAlert } from 'components/Alert/ApiErrorMessage';
import { LoadingMessage } from 'components/Loader/LoadingMessage';
import { Alert } from 'components/Alert/Alert';
import { ApiError } from 'interfaces/api';
import { Pagination } from 'components/Pagination/Pagination';
import PageCartoUserItem from './PageCartoUserItem';
import { ContentType, QueryParams, strapi } from 'services/strapi';
import { PageCarto } from 'interfaces/page-carto';

export const PageCartoUserList = ({ nameInput, tagsTable }: any) => {
  const paginationLimit = 9;

  const [page, setPage] = useState(1);
  const [currentUser, setCurrentUser] = useState('');

  useEffect(() => {
    const userGetData = async () => {
      return strapi.token && (await strapi.currentUser(strapi.token));
    };
    userGetData().then((user) => {
      if (user) {
        setCurrentUser(user.username);
      }
    });
  }, []);

  const {
    data: response,
    isError,
    error,
    isLoading,
  } = useQuery({
    queryKey: ['page-carto-user', page, currentUser, nameInput, tagsTable],
    queryFn: () => {
      return strapi.get<PageCarto>(ContentType.PAGE_CARTOS, {
        filters: {
          owner: {
            username: {
              $eq: currentUser,
            },
          },
          $or: [
            {
              name:
                nameInput != ''
                  ? {
                      $contains: nameInput,
                    }
                  : {},
            },
            {
              html:
                nameInput != ''
                  ? {
                      $contains: nameInput,
                    }
                  : {},
            },
          ],
          $and: [
            {
              tags: {
                name:
                  tagsTable.length != 0
                    ? {
                        $in: tagsTable,
                      }
                    : {},
              },
            },
          ],
        } as QueryParams['filters'],
        populate: '*',
        sort: 'createdAt:desc',
        pagination: {
          page,
          pageSize: paginationLimit,
        },
      });
    },
  });

  if (isLoading) {
    return <LoadingMessage />;
  }

  if (isError) {
    return <ApiErrorAlert error={error as ApiError} />;
  }

  if (!response || response.data.length === 0) {
    return <Alert type="info">Aucun contenu Ã  afficher.</Alert>;
  }

  const { data, meta } = response;

  return (
    <div>
      <div className="grid grid-cols-3 gap-y-10 gap-x-6">
        {data.map((page) => {
          return <PageCartoUserItem key={page.id} {...page} />;
        })}
      </div>
      <div className="flex flex-row mt-5 justify-center">
        <Pagination
          page={page}
          onPaginateNext={() => setPage(page + 1)}
          onPaginatePrevious={() => setPage(page - 1)}
          onPaginate={(p: number) => setPage(p)}
          totalPages={meta.pagination.pageCount || 0}
        />
      </div>
    </div>
  );
};
