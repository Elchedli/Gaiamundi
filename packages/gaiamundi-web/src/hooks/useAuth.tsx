import React from 'react';
import {
  QueryObserverResult,
  RefetchOptions,
  UseMutateFunction,
  useQuery,
  useQueryClient,
} from 'react-query';

import { Alert } from 'components/Alert/Alert';
import { LoadingMessage } from 'components/Loader/LoadingMessage';
import { User } from 'interfaces/user';
import { strapi } from '../services/strapi';
import { useLocalStorageState } from './useLocalStorageState';

// const updateAccount = async (id: number, userData: UserSignUpFields) => {
//   return (await strapi.update(ContentType.USERS, id, userData)) as User;
// };

export interface AuthContextValue {
  user: User | undefined;
  isAuthenticated: boolean;
  setUser: (data: User | undefined) => void;
  authenticate: (user: User, jwt: string) => void;
  logout: UseMutateFunction;
  refetchUser: (
    options?: RefetchOptions | undefined
  ) => Promise<QueryObserverResult<User, Error>>;
  error: Error | null;
}

const AuthContext = React.createContext<AuthContextValue | null>(null);
AuthContext.displayName = 'AuthContext';

export interface AuthProviderProps {
  children: React.ReactNode;
}

export const AuthProvider = ({ children }: AuthProviderProps): JSX.Element => {
  const { value: jwtToken, persistValue: persistJwtToken } =
    useLocalStorageState('token');
  const queryClient = useQueryClient();
  const key = 'auth-user';

  const {
    data: user,
    error,
    isLoading,
    refetch,
  } = useQuery<User, Error>({
    queryKey: key,
    queryFn: async () => {
      return await strapi.currentUser(jwtToken);
    },
    enabled: !!jwtToken,
  });

  const setUser = (data: User | undefined) =>
    queryClient.setQueryData(key, data);

  const authenticate = (user: User, jwt: string) => {
    setUser(user);
    persistJwtToken(jwt);
  };

  const logout = () => {
    persistJwtToken('');
    queryClient.clear();
  };

  if (isLoading) {
    return <LoadingMessage />;
  }

  return (
    <AuthContext.Provider
      value={{
        user,
        isAuthenticated: !!jwtToken,
        error,
        setUser,
        authenticate,
        refetchUser: refetch,
        logout,
      }}
    >
      {error && <Alert>{error.message}</Alert>}
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = React.useContext(AuthContext);
  if (!context) {
    throw new Error(`useAuth must be used within an AuthProvider`);
  }
  return context;
};
