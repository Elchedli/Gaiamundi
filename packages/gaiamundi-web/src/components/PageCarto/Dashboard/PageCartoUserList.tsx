import { Alert } from 'components/Alert/Alert';
import { ApiErrorAlert } from 'components/Alert/ApiErrorMessage';
import { LoadingMessage } from 'components/Loader/LoadingMessage';
import PageCartoItem from 'components/PageCarto/List/PageCartoItem';
import { Pagination } from 'components/Pagination/Pagination';
import { useAuth } from 'hooks/useAuth';
import { ApiError } from 'interfaces/api';
import { useState } from 'react';
import { useQuery } from 'react-query';
import { getPageCartoByTagsAndSearch } from 'services/page-carto';
export const PageCartoUserList = ({
  searchKeywords,
  selectedTags,
}: {
  searchKeywords: string;
  selectedTags: number[];
}) => {
  const [page, setPage] = useState(1);
  const { user } = useAuth();
  const {
    data: response,
    isError,
    error,
    isLoading,
  } = useQuery({
    queryKey: ['page-carto-user', page, searchKeywords, selectedTags],
    queryFn: async () =>
      getPageCartoByTagsAndSearch(page, searchKeywords, selectedTags, user),
  });
  if (isLoading) {
    return (
      <LoadingMessage data-testid="page-carto-user-list-loading-message" />
    );
  }
  if (isError) {
    return (
      <ApiErrorAlert
        error={error as ApiError}
        data-testid="pagecarto-user-list-error-message"
      />
    );
  }
  return (
    <div>
      {!response || response.data.length === 0 ? (
        <Alert
          type="info"
          className="bg-transparent border-transparent w-full grid justify-center items-center"
        >
          <div data-testid="pagecarto-user-list-error-message">
            Aucun contenu Ã  afficher.
          </div>
        </Alert>
      ) : (
        <>
          <div className="grid grid-cols-3 gap-y-10 gap-x-6">
            {response.data.map((page) => {
              return <PageCartoItem key={page.id} {...page} />;
            })}
          </div>
          <div className="flex flex-row mt-5 justify-center">
            <Pagination
              page={page}
              onPaginate={(p: number) => setPage(p)}
              totalPages={response?.meta.pagination.pageCount || 0}
            />
          </div>
        </>
      )}
    </div>
  );
};
