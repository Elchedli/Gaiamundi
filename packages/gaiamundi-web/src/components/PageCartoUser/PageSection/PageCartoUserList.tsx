import { useQuery } from 'react-query';
import { useState } from 'react';
import { ApiErrorAlert } from 'components/Alert/ApiErrorMessage';
import { LoadingMessage } from 'components/Loader/LoadingMessage';
import { Alert } from 'components/Alert/Alert';
import { ApiError } from 'interfaces/api';
import { Pagination } from 'components/Pagination/Pagination';
import { ContentType, QueryParams, strapi } from 'services/strapi';
import { PageCarto } from 'interfaces/page-carto';
import { useAuth } from 'hooks/useAuth';
import PageCartoItem from 'components/PageCarto/PageCartoItem';
import { useFilterPageCarto } from 'hooks/useFilter';

export const PageCartoUserList = () => {
  const paginationLimit = 9;
  const [page, setPage] = useState(1);
  const { user } = useAuth();
  const { state } = useFilterPageCarto();

  const { nameInput, tagsSelected } = state;
  const tagsSelectedName = tagsSelected.map((a) => a.attributes.name);
  const {
    data: response,
    isError,
    error,
    isLoading,
  } = useQuery({
    queryKey: ['page-carto-user', page, state.nameInput, state.tagsSelected],
    queryFn: () => {
      return strapi.get<PageCarto>(ContentType.PAGE_CARTOS, {
        filters: {
          owner: {
            id: {
              $eq: user?.id,
            },
          },
          $or: [
            {
              name:
                nameInput != ''
                  ? {
                      $contains: nameInput,
                    }
                  : {},
            },
            {
              html:
                nameInput != ''
                  ? {
                      $contains: nameInput,
                    }
                  : {},
            },
          ],
          $and: [
            {
              tags: {
                name:
                  tagsSelectedName.length != 0
                    ? {
                        $in: tagsSelectedName,
                      }
                    : {},
              },
            },
          ],
        } as QueryParams['filters'],
        populate: '*',
        sort: 'createdAt:desc',
        pagination: {
          page,
          pageSize: paginationLimit,
        },
      });
    },
  });

  if (isLoading) {
    return <LoadingMessage />;
  }

  if (isError) {
    return <ApiErrorAlert error={error as ApiError} />;
  }

  if (!response || response.data.length === 0) {
    return <Alert type="info">Aucun contenu Ã  afficher.</Alert>;
  }

  const { data, meta } = response;

  return (
    <div>
      <div className="grid grid-cols-3 gap-y-10 gap-x-6">
        {data.map((page) => {
          return <PageCartoItem key={page.id} {...page} />;
        })}
      </div>
      <div className="flex flex-row mt-5 justify-center">
        <Pagination
          page={page}
          onPaginateNext={() => setPage(page + 1)}
          onPaginatePrevious={() => setPage(page - 1)}
          onPaginate={(p: number) => setPage(p)}
          totalPages={meta.pagination.pageCount || 0}
        />
      </div>
    </div>
  );
};
