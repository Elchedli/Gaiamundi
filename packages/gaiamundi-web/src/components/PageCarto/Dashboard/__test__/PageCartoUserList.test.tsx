import { fireEvent, render } from '@testing-library/react';
import { useAuth } from 'hooks/useAuth';
import { ApiCollection } from 'interfaces/api';
import { PageCarto } from 'interfaces/page-carto';
import { useQuery } from 'react-query';
import { MemoryRouter as Router } from 'react-router-dom';
import { mockPageCartoData, mockUser } from 'utils/mocks/data';
import { PageCartoUserList } from '../PageCartoUserList';

jest.mock('hooks/useAuth');
jest.mock('react-query');

describe('PageCartoUserList', () => {
  beforeEach(() => {
    (useAuth as jest.Mock).mockImplementation(() => ({
      isAuthenticated: true,
      user: mockUser,
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders loading message when loading', () => {
    (useQuery as jest.Mock).mockReturnValueOnce({
      data: null,
      isError: false,
      error: null,
      isLoading: true,
    });

    const { container, getByTestId } = render(
      <PageCartoUserList searchKeywords="" selectedTags={[]} />
    );

    expect(getByTestId('loading-message')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders error message when there is an error', () => {
    const mockError = {
      message: 'An error occurred',
      statusCode: 500,
    };

    (useQuery as jest.Mock).mockReturnValueOnce({
      data: null,
      isError: true,
      error: mockError,
      isLoading: false,
    });

    const { container, getByText } = render(
      <PageCartoUserList searchKeywords="" selectedTags={[]} />
    );

    expect(getByText(mockError.message)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders message when there is no data', () => {
    (useQuery as jest.Mock).mockReturnValueOnce({
      data: {
        data: [],
        meta: {
          pagination: {
            pageCount: 0,
          },
        },
      },
      isError: false,
      error: null,
      isLoading: false,
    });

    const { container, getByTestId } = render(
      <PageCartoUserList searchKeywords="" selectedTags={[]} />
    );

    expect(getByTestId('error-message')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders pages when there is data', () => {
    const pageCartos: ApiCollection<PageCarto> = {
      data: Array.from({ length: 2 }, (_, index) => ({
        ...mockPageCartoData,
        id: index + 1,
        name: `Test pageCarto ${index + 1}`,
      })),
      meta: {
        pagination: {
          pageCount: 2,
          total: 2,
        },
      },
    };

    (useQuery as jest.Mock).mockImplementation(() => ({
      data: pageCartos,
    }));

    const { container, getByText, getByTestId } = render(
      <Router>
        <PageCartoUserList searchKeywords="" selectedTags={[]} />
      </Router>
    );

    const pagination = getByTestId('pagination');

    pageCartos.data.forEach((pageCarto) => {
      expect(getByText(pageCarto.name)).toBeInTheDocument();
    });
    expect(pagination.children).toHaveLength(3);
    expect(pagination.children[0].textContent).toEqual('1');
    expect(pagination.children[1].textContent).toEqual('2');

    fireEvent.click(pagination.children[1]);

    expect((useQuery as jest.Mock).mock.calls[0][0].queryKey[1]).toBe(1);
    expect(container).toMatchSnapshot();
  });
});
