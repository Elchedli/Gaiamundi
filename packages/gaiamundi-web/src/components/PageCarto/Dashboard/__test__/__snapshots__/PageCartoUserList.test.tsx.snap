// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageCartoUserList renders "Aucun contenu à afficher" message when there is no data 1`] = `
<div>
  <div>
    <div
      class="p-3 my-3 rounded border border-blue-500 bg-blue-200 text-blue-700 bg-transparent border-transparent w-fit grid justify-center items-center"
      role="alert"
    >
      <div
        class="flex flex-row"
      >
        <div>
          <div
            data-testid="error-message"
          >
            Aucun contenu à afficher.
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PageCartoUserList renders error message when there is an error 1`] = `
<div>
  <div
    class="p-3 my-3 rounded border border-red-500 bg-red-200 text-red-700"
    role="alert"
  >
    <div
      class="flex flex-row"
    >
      <svg
        aria-hidden="true"
        class="mr-2"
        fill="currentColor"
        height="24"
        stroke="current"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.47 5.47a.75.75 0 011.06 0L12 10.94l5.47-5.47a.75.75 0 111.06 1.06L13.06 12l5.47 5.47a.75.75 0 11-1.06 1.06L12 13.06l-5.47 5.47a.75.75 0 01-1.06-1.06L10.94 12 5.47 6.53a.75.75 0 010-1.06z"
          fill-rule="evenodd"
        />
      </svg>
      <div>
        An error occurred
      </div>
    </div>
  </div>
</div>
`;

exports[`PageCartoUserList renders loading message when loading 1`] = `
<div>
  <div
    class="flex grow justify-center items-center"
    data-testid="loading-message"
  >
    <div
      class="flex flex-col"
    >
      <svg
        class="m-auto"
        fill="#88CDD3"
        viewBox="0 0 40 40"
        width="200"
      >
        <path
          d="M20.201,5.169c-8.254,0-14.946,6.692-14.946,14.946c0,8.255,6.692,14.946,14.946,14.946 s14.946-6.691,14.946-14.946C35.146,11.861,28.455,5.169,20.201,5.169z M20.201,31.749c-6.425,0-11.634-5.208-11.634-11.634 c0-6.425,5.209-11.634,11.634-11.634c6.425,0,11.633,5.209,11.633,11.634C31.834,26.541,26.626,31.749,20.201,31.749z"
          opacity="0.5"
        />
        <path
          d="M26.013,10.047l1.654-2.866c-2.198-1.272-4.743-2.012-7.466-2.012h0v3.312h0 C22.32,8.481,24.301,9.057,26.013,10.047z"
        >
          <animatetransform
            attributeName="transform"
            attributeType="xml"
            dur="0.8s"
            from="0 20 20"
            repeatCount="indefinite"
            to="360 20 20"
            type="rotate"
          />
        </path>
      </svg>
      <h1
        class="text-center text-xl font-semibold"
      >
        Loading ...
      </h1>
    </div>
  </div>
</div>
`;

exports[`PageCartoUserList renders message when there is no data 1`] = `
<div>
  <div>
    <div
      class="p-3 my-3 rounded border border-blue-500 bg-blue-200 text-blue-700 bg-transparent border-transparent w-fit grid justify-center items-center"
      role="alert"
    >
      <div
        class="flex flex-row"
      >
        <div>
          <div
            data-testid="error-message"
          >
            Aucun contenu à afficher.
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PageCartoUserList renders pages when there is data 1`] = `
<div>
  <div>
    <div
      class="grid grid-cols-3 gap-y-10 gap-x-6"
    >
      <div
        class="max-w-sm rounded overflow-hidden shadow-lg"
      >
        <a
          href="/page-carto/1"
        >
          <div
            class="aspect-w-1 aspect-h-1 w-full overflow-hidden bg-gray-400 xl:aspect-w-7 xl:aspect-h-8"
          >
            <img
              class="h-48 w-full object-cover object-center group-hover:opacity-75"
              src="http://localhost:1337/api/v1/geo-maps/id/thumbnailImage?720"
            />
          </div>
          <div
            class="p-3"
          >
            <h2>
              Test pageCarto 1
            </h2>
            <div
              class="text-gray-500"
            >
              <div
                class="text-sm my-1"
              >
                <div
                  class="my-1 font-bold"
                >
                  Carte : 
                  Test Map
                </div>
                <p>
                  Test HTML
                </p>
                <div
                  class="text-xs mt-2"
                >
                  Créer par : 
                  testuser
                </div>
              </div>
              <div
                class="mt-3"
              >
                <a
                  href="#"
                >
                  <span
                    class="bg-gray-200 text-gray-700 rounded-full px-3 py-1 text-sm font-semibold mr-2 mb-2 inline-block"
                  >
                    <span>
                      tag1
                    </span>
                  </span>
                </a>
                <a
                  href="#"
                >
                  <span
                    class="bg-gray-200 text-gray-700 rounded-full px-3 py-1 text-sm font-semibold mr-2 mb-2 inline-block"
                  >
                    <span>
                      tag2
                    </span>
                  </span>
                </a>
              </div>
            </div>
          </div>
        </a>
      </div>
      <div
        class="max-w-sm rounded overflow-hidden shadow-lg"
      >
        <a
          href="/page-carto/2"
        >
          <div
            class="aspect-w-1 aspect-h-1 w-full overflow-hidden bg-gray-400 xl:aspect-w-7 xl:aspect-h-8"
          >
            <img
              class="h-48 w-full object-cover object-center group-hover:opacity-75"
              src="http://localhost:1337/api/v1/geo-maps/id/thumbnailImage?720"
            />
          </div>
          <div
            class="p-3"
          >
            <h2>
              Test pageCarto 2
            </h2>
            <div
              class="text-gray-500"
            >
              <div
                class="text-sm my-1"
              >
                <div
                  class="my-1 font-bold"
                >
                  Carte : 
                  Test Map
                </div>
                <p>
                  Test HTML
                </p>
                <div
                  class="text-xs mt-2"
                >
                  Créer par : 
                  testuser
                </div>
              </div>
              <div
                class="mt-3"
              >
                <a
                  href="#"
                >
                  <span
                    class="bg-gray-200 text-gray-700 rounded-full px-3 py-1 text-sm font-semibold mr-2 mb-2 inline-block"
                  >
                    <span>
                      tag1
                    </span>
                  </span>
                </a>
                <a
                  href="#"
                >
                  <span
                    class="bg-gray-200 text-gray-700 rounded-full px-3 py-1 text-sm font-semibold mr-2 mb-2 inline-block"
                  >
                    <span>
                      tag2
                    </span>
                  </span>
                </a>
              </div>
            </div>
          </div>
        </a>
      </div>
    </div>
    <div
      class="flex flex-row mt-5 justify-center"
    >
      <div
        class="flex flex-row mt-5"
        data-testid="pagination"
      >
        <button
          class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600 mr-3"
          data-testid="button-element"
          type="button"
        >
          <span
            class="flex items-center text-sm px-4 py-2 rounded-md"
          >
            &lt; Précédent
          </span>
        </button>
        <button
          class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600 mx-2"
          data-testid="button-element"
          type="button"
        >
          <span
            class="flex items-center text-sm px-4 py-2 rounded-md"
          >
            1
          </span>
        </button>
        <button
          class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600 cursor-not-allowed opacity-50 bg-gray-200"
          data-testid="button-element"
          disabled=""
          type="button"
        >
          <span
            class="flex items-center text-sm px-4 py-2 rounded-md"
          >
            2
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
